xbmc/windowing/gbm/DRMAtomic.cpp

```diff
@@ -23,6 +25,57 @@
 
 using namespace KODI::WINDOWING::GBM;
 
+const std::string CDRMAtomic::SETTING_VIDEOSCREEN_HW_SCALING_FILTER = "videoscreen.hwscalingfilter";
+
+enum drm_scaling_filter
+{
+  DRM_SCALING_FILTER_DEFAULT,
+  DRM_SCALING_FILTER_NEAREST_NEIGHBOR,
```

Comment:
This incorrectly hard-codes the values of the DRM enum here.

The uAPI doesn't have this guarantee, e.g. "Nearest Neighbor" could be exposed as the integer 5 in the future.

Reply:
We had discussed that point [here before ](https://github.com/xbmc/xbmc/pull/18194#discussion_r457708467), In fact it was my initial implementation to read the DRM properties and set the proper value as per request. But later concluded on this implementation based on the discussions in this PR.

Reply:
The current AddProperty API does not take/pulls the valid values from DRM property, so went with the current implement with enums.
But if you think the code below make sense, we can incorporate the same instead.
`bool CDRMAtomic::AddPropertyValueByName(struct drm_object* object, const char* name, const char* type)
{
  for (uint32_t i = 0; i < object->props->count_props; i++)
  {
    if (!strcmp(object->props_info[i]->name, name))
    {
      for (uint32_t j = 0; j < object->props_info[i]->count_enums; j++)
      {
        if (!strcmp(object->props_info[i]->enums[j].name, type))
        {
          if (drmModeAtomicAddProperty(m_req, object->id, property_id,
                                       m_gui_plane->props_info[i]->enums[j].value) < 0)
          {
                CLog::Log(LOGERROR, "CDRMAtomic::%s - could not add property %s", __FUNCTION__, name);
                return false;
          }
        }
      }
    }
  }
  return true;
}
`
This was the exact implementation which I had before.
Please let me know your thought on the same.

Reply:
This code is very hard to read, please put it inside triple-backquotes like so:

````
```cpp
<insert code here>
```
````

Reply:
```cpp
bool CDRMAtomic::AddPropertyValueByName(struct drm_object* object, const char* name, const char* type)
{
  for (uint32_t i = 0; i < object->props->count_props; i++)
  {
    if (!strcmp(object->props_info[i]->name, name))
    {
      for (uint32_t j = 0; j < object->props_info[i]->count_enums; j++)
      {
        if (!strcmp(object->props_info[i]->enums[j].name, type))
        {
          if (drmModeAtomicAddProperty(m_req, object->id, property_id,
                                       m_gui_plane->props_info[i]->enums[j].value) < 0)
          {
            CLog::Log(LOGERROR, "CDRMAtomic::%s - could not add property %s", __FUNCTION__, name);
            return false;
          }
        }
      }
    }
  }
  return true;
}
```

Reply:
We do have `bool SupportsPropertyAndValue(const char* name, uint64_t value);` see https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/gbm/drm/DRMObject.h#L37

Reply:
@slattann This code doesn't fail when the prop or its value is missing. Could also use some clean-up to avoid having an intentation which is this deep. But the idea is good.

@lrusak The problem is that you can't just hard-code an `uint64_t` in your program. You need to get the `uint64_t` from the kernel.

Reply:
> @lrusak The problem is that you can't just hard-code an `uint64_t` in your program. You need to get the `uint64_t` from the kernel.

@emersion I'm not sure I follow. I assume that the enum would be in the kernel UAPI when the patches merge?

Reply:
No, the enum value is not in the uAPI. The enum value is exposed by the kernel at runtime via `drmModePropertyRes`. The enum value may change from one driver to another, and from one plane to another.

Reply:
> No, the enum value is not in the uAPI. The enum value is exposed by the kernel at runtime via `drmModePropertyRes`. The enum value may change from one driver to another, and from one plane to another.
> 
> Only the enum name (the string) is in the uAPI.

That's no problem, we evaluate all properties for a drm object at startup.

Reply:
This issue is this enum. The value of `DRM_SCALING_FILTER_NEAREST_NEIGHBOR` is hard-coded to 1, returned from `CDRMAtomic::GetScalingFilterType`, then passed as-is to `AddProperty` in `CDRMAtomic::SetScalingFilter`.

Reply:
> The issue is this enum. The value of `DRM_SCALING_FILTER_NEAREST_NEIGHBOR` is hard-coded to 1, returned from `CDRMAtomic::GetScalingFilterType`, then passed as-is to `AddProperty` in `CDRMAtomic::SetScalingFilter`.

it can be fixed :smiley_cat:

Reply:
It must be fixed. @emersion is right.

This is actually one of those kinds of bugs that could lead to problems for kernel developers later, if they have to work around broken userspace assumptions to keep broken userspace working.

The general rule indeed is that KMS property enum values MUST be recognized by their string names, not by their numeric values. Unfortunately the KMS UAPI has past mistakes where the numerical values are already in the UAPI headers - that cannot be fixed, but all new KMS properties need to follow the rules, hence userspace can't pick an enum value by number. That is the reason you also cannot find the numbers in the kernel/libdrm UAPI headers.

Reply:
I have created another PR for the fix: https://github.com/xbmc/xbmc/pull/18567
Kindly let me know if that is fine and we can go ahead w/ this PR and get Kernel patches reviewed and merged ?